### PR TITLE
PXC 4050 PS 8.0.31 merge

### PR DESCRIPTION
**Pull request by VarunNagaraju**

### CODE REVIEW

File: mysql-test/include/mtr_check.sql [hunk 1]
```diff
@@ -139,16 +133,10 @@ BEGIN
 
   -- Dump all databases, there should be none
   -- except those that was created during bootstrap
-<<<<<<< HEAD
   -- and the mtr_wsrep_notify schema which is populated by the std_data/wsrep_notify.sh script
   -- and the suite/galera/t/galera_var_notify_cmd.test
   -- and the wsrep_schema schema that may be created by Galera
-  SELECT * FROM INFORMATION_SCHEMA.SCHEMATA WHERE SCHEMA_NAME NOT IN ('mtr_wsrep_notify', 'wsrep_schema') ORDER BY SCHEMA_NAME;
```
Comment (venkatesh-prasad-v) [inline mysql-test/include/mtr_check.sql:139]: The conflict resolution is not considerng the clause `SCHEMA_NAME NOT IN ('mtr_wsrep_notify', 'wsrep_schema')`

File: mysql-test/r/all_persisted_variables.result [hunk 2]
```diff
@@ -35,17 +35,10 @@ CREATE TABLE all_vars (id INT PRIMARY KEY AUTO_INCREMENT, var_name VARCHAR(64),
 INSERT INTO all_vars (var_name, var_value)
 SELECT * FROM performance_schema.global_variables
 WHERE variable_name NOT IN
-<<<<<<< HEAD
 ('innodb_default_encryption_key_id',
 'rbr_exec_mode',
 'pxc_maint_mode',
 'pxc_strict_mode')
-||||||| merged common ancestors
-('innodb_default_encryption_key_id',
```
Comment (venkatesh-prasad-v) [inline mysql-test/r/all_persisted_variables.result]: `innodb_default_encryption_key_id` has been removed in PS, but it is not removed in PXC. Please check this.

File: mysql-test/t/log_buffered-big.test [hunk 3]
```diff
@@ -49,13 +49,7 @@ SET PERSIST log_error_suppression_list="ER_BASEDIR_SET_TO,ER_DD_INIT_FAILED";
 # moment it will not be able to do so as InnoDB doublewrite component will not
 # be properly shut down by the first (original) instance.
 --let LOG_FILE1= $MYSQLTEST_VARDIR/tmp/test1.err
-<<<<<<< HEAD
---let restart_parameters="restart: --skip-innodb-doublewrite --log-error=$LOG_FILE_TRAD --log-error-services=$log_services --log-error-suppression-list=$log_sup_list"
-||||||| merged common ancestors
---let restart_parameters="restart: --log-error=$LOG_FILE_TRAD --log-error-services=$log_services --log-error-suppression-list=$log_sup_list"
-=======
 --let restart_parameters="restart: --log-error=$LOG_FILE_TRAD --log-error-verbosity=3"
->>>>>>> Percona-Server-8.0.31-23
```
Comment (venkatesh-prasad-v) [inline mysql-test/t/log_buffered-big.test]: Why was `--skip-innodb-doublewrite` omitted?

File: packaging/deb-in/deb_debug.cmake [hunk 4]
```diff
@@ -117,14 +117,6 @@ usr/lib/mysql/plugin/debug/component_test_udf_registration.so
 usr/lib/mysql/plugin/debug/component_test_host_application_signal.so
 usr/lib/mysql/plugin/debug/component_test_mysql_current_thread_reader.so
 usr/lib/mysql/plugin/debug/component_test_mysql_runtime_error.so
-<<<<<<< HEAD
-||||||| merged common ancestors
-usr/lib/mysql/plugin/debug/component_test_component_deinit.so
-=======
-usr/lib/mysql/plugin/debug/component_test_component_deinit.so
-usr/lib/mysql/plugin/debug/component_test_mysql_command_services.so
-usr/lib/mysql/plugin/debug/test_services_command_services.so
```
Comment (venkatesh-prasad-v) [inline packaging/deb-in/deb_debug.cmake]: Any reasons why `component_test_component_deinit.so` was omitted in the first place? Can we have PXC inherit new changes as well?

File: packaging/deb-in/mysql-packagesource-test.install.in [hunk 5]
```diff
@@ -47,14 +47,6 @@ usr/lib/mysql/plugin/component_test_udf_registration.so
 usr/lib/mysql/plugin/component_test_host_application_signal.so
 usr/lib/mysql/plugin/component_test_mysql_current_thread_reader.so
 usr/lib/mysql/plugin/component_test_mysql_runtime_error.so
-<<<<<<< HEAD
-||||||| merged common ancestors
-usr/lib/mysql/plugin/component_test_component_deinit.so
-=======
-usr/lib/mysql/plugin/component_test_component_deinit.so
-usr/lib/mysql/plugin/component_test_mysql_command_services.so
-usr/lib/mysql/plugin/test_services_command_services.so
```
Comment (venkatesh-prasad-v) [inline packaging/deb-in/mysql-packagesource-test.install.in]: Same as above

File: sql/sql_admin.h [hunk 6]
```diff
@@ -360,17 +360,11 @@ class Sql_cmd_alter_user_default_role : public Sql_cmd {
 
 enum alter_instance_action_enum {
   ROTATE_INNODB_MASTER_KEY,
-<<<<<<< HEAD
   ROTATE_INNODB_SYSTEM_KEY,
   ROTATE_REDO_SYSTEM_KEY,
 #ifdef WITH_WSREP
   ROTATE_GCACHE_MASTER_KEY,
 #endif
-||||||| merged common ancestors
```
Comment (venkatesh-prasad-v) [inline sql/sql_admin.h]: ROTATE_INNODB_SYSTEM_KEY and ROTATE_REDO_SYSTEM_KEY must be removed from PXC sources as well.

File: sql/sql_yacc.yy [hunk 7]
```diff
@@ -16380,40 +16379,6 @@ alter_instance_action:
...
-            }
-          }
-          | ROTATE_SYM ident_or_text SYSTEM_SYM KEY_SYM
-          {
-            if (is_identifier($2, "REDO"))
-            {
-              $$= NEW_PTN PT_alter_instance(ROTATE_REDO_SYSTEM_KEY, EMPTY_CSTR, 0);
-=======
-              $$= NEW_PTN PT_alter_instance(ROTATE_BINLOG_MASTER_KEY, EMPTY_CSTR);
->>>>>>> Percona-Server-8.0.31-23
```
Comment (venkatesh-prasad-v) [inline sql/sql_yacc.yy]: Please remove support for ROTATE_INNODB_SYSTEM_KEY and ROTATE_REDO_SYSTEM_KEY in parser.

File: storage/innobase/srv/srv0start.cc [hunk 8]
```diff
@@ -205,18 +205,12 @@ mysql_pfs_key_t srv_ts_alter_encrypt_thread_key;
...
 
-<<<<<<< HEAD
 int unlock_keyrings(THD *thd);
 
 #ifdef WITH_WSREP
 extern bool wsrep_recovery;
 #endif /* WITH_WSREP */
 
-||||||| merged common ancestors
-int unlock_keyrings(THD *thd);
```
Comment (venkatesh-prasad-v) [inline storage/innobase/srv/srv0start.cc]: If PS has removed this function, then it must be removed in PXC as well.

File: sql/sql_alter_instance.cc [hunk 9]
```diff
@@ -186,7 +186,15 @@ bool Rotate_innodb_master_key::execute() {
 */
 extern bool wsrep_rotate_master_key();
 bool Rotate_gcache_master_key::execute() {
-  if (check_security_context()) return true;
+
+  Security_context *sctx = m_thd->security_context();
+  if (!sctx->check_access(SUPER_ACL) &&
+      !sctx->has_global_grant(STRING_WITH_LEN("BINLOG_ENCRYPTION_ADMIN"))
```
Comment (venkatesh-prasad-v) [inline sql/sql_alter_instance.cc:192]: Why `BINLOG_ENCRYPTION_ADMIN`? @kamil-holubicki Can you please confirm if we need `BINLOG_ENCRYPTION_ADMIN` privilege to enable gcache encryption.

**Review by venkatesh-prasad-v — CHANGES_REQUESTED** [review]

File: mysql-test/include/mtr_check.sql (hunk 1 above)
Comment (VarunNagaraju) [reply mysql-test/include/mtr_check.sql:139]: Fixed.

File: mysql-test/r/all_persisted_variables.result (hunk 2 above)
Comment (VarunNagaraju) [reply mysql-test/r/all_persisted_variables.result]: It has been completely removed. The references fixed in a subsequent commit.

File: mysql-test/t/log_buffered-big.test (hunk 3 above)
Comment (VarunNagaraju) [reply mysql-test/t/log_buffered-big.test]: My bad. Fixed the conflict resolution.

File: packaging/deb-in/deb_debug.cmake (hunk 4 above)
Comment (VarunNagaraju) [reply packaging/deb-in/deb_debug.cmake]: Done.

File: packaging/deb-in/mysql-packagesource-test.install.in (hunk 5 above)
Comment (VarunNagaraju) [reply packaging/deb-in/mysql-packagesource-test.install.in]: Done.

File: sql/sql_admin.h (hunk 6 above)
Comment (VarunNagaraju) [reply sql/sql_admin.h]: Done.

File: sql/sql_yacc.yy (hunk 7 above)
Comment (VarunNagaraju) [reply sql/sql_yacc.yy]: Done.

File: storage/innobase/srv/srv0start.cc (hunk 8 above)
Comment (VarunNagaraju) [reply storage/innobase/srv/srv0start.cc]: Done.

File: packaging/deb-in/deb_debug.cmake (hunk 4 above)
Comment (kamil-holubicki) [reply packaging/deb-in/deb_debug.cmake]: packaging directory is not used, however I agree, we should align with upstream to avoid merge conflicts in the future

File: storage/innobase/handler/ha_innodb.cc [hunk 10]
```diff
@@ -5684,31 +5644,15 @@ static int innodb_init(void *p) {
...
 #endif /* WITH_WSREP */
-||||||| merged common ancestors
-  innobase_hton->flags = HTON_SUPPORTS_EXTENDED_KEYS |
-                         HTON_SUPPORTS_FOREIGN_KEYS | HTON_SUPPORTS_ATOMIC_DDL |
-                         HTON_CAN_RECREATE | HTON_SUPPORTS_SECONDARY_ENGINE |
-                         HTON_SUPPORTS_TABLE_ENCRYPTION |
-                         HTON_SUPPORTS_ONLINE_BACKUPS | HTON_SUPPORTS_COMPRESSED_COLUMNS |
-                         HTON_SUPPORTS_GENERATED_INVISIBLE_PK;
-=======
   innobase_hton->flags =
```
Comment (kamil-holubicki) [inline storage/innobase/handler/ha_innodb.cc:5647]: This will overwrite the flag set above in ifdef WITH_WSREP.
WITH_WSREP block should go after this assignment.

File: sql/sql_alter_instance.cc (hunk 9 above)
Comment (kamil-holubicki) [reply sql/sql_alter_instance.cc:192]: It has to be the same as in `Rotate_innodb_master_key::execute()`, so we need `ENCRYPTION_KEY_ADMIN`

File: scripts/mysql_system_tables_fix.sql [hunk 11]
```diff
@@ -803,9 +803,9 @@ COMMIT;
 
 -- Add the privilege AUTHENTICATION_POLICY_ADMIN for every user who has the SYSTEM_VARIABLES_ADMIN privilege
 -- provided that there isn't a user who already has the privilege AUTHENTICATION_POLICY_ADMIN.
-SET @hadAuthenticationPolicyAdminPriv = (SELECT COUNT(*) FROM global_grants WHERE priv = 'AUTHENTICATION_POLICY_ADMIN');
+SET @hadAuthenticationPolicyAdminPriv = (SELECT COUNT(*) FROM global_grants WHERE priv = 'AUTHENTICATION_POLICY_ADMIN' AND user NOT IN ('mysql.session'));
```
Comment (kamil-holubicki) [inline scripts/mysql_system_tables_fix.sql]: Why this change?

**Review by kamil-holubicki — CHANGES_REQUESTED** [review]
Please check if everything is fine with i_s_schema_definition_debug.test. It has 2 versions for CS and CI filesystems.

File: cmake/package_name.cmake [hunk 12]
```diff
@@ -161,16 +161,8 @@ IF(NOT CPACK_PACKAGE_FILE_NAME)
 ENDIF()
 
 IF(NOT CPACK_SOURCE_PACKAGE_FILE_NAME)
-<<<<<<< HEAD
   SET(CPACK_SOURCE_PACKAGE_FILE_NAME "Percona-XtraDB-Cluster-${VERSION}")
   IF(WITH_NDBCLUSTER)
```
Comment (dutow) [inline cmake/package_name.cmake]: It probably doesn't matter for pxc, but this should be WITH_NDB

File: scripts/mysql_system_tables_fix.sql (hunk 11 above)
Comment (VarunNagaraju) [reply scripts/mysql_system_tables_fix.sql]: @kamil-holubicki This is the change I mentioned when I did the preliminary merge as well.
The upstream bug which introduced the new change _Bug #33165861: WL#14183 new privileges not taken into account for upgrades_ and the comment says that if there isn't a user who already has AUTHENTICATION_POLICY_ADMIN privilege, the new privilege upgrade happens. But looking at the script, it seems like mysql.session user is NOT considered whenever privilege upgrade like these are made. So, I think mysql.session user having AUTHENTICATION_POLICY_ADMIN privilege should also NOT be considered here. What do you think? Should we report an upstream bug for it?
